drivers/sgp30/sgp30.c

```diff
@@ -0,0 +1,353 @@
+    SGP30_CMD_READ_SERIAL               = 0x3682,   /**< Read serial number */
+    SGP30_CMD_SOFT_RESET                = 0x0006,   /**< Perform General Call reset */
+} sgp30_cmd_t;
+/** @} */
+
+/**
+ * @name      Max delays for SGP30 I2C commands completion
+ * @{
+ */
+#define SGP30_DELAY_INIT_AIR_QUALITY        (10 * MS_PER_SEC)
+#define SGP30_DELAY_MEASURE_AIR_QUALITY     (12 * MS_PER_SEC)
+#define SGP30_DELAY_GET_BASELINE            (10 * MS_PER_SEC)
+#define SGP30_DELAY_SET_BASELINE            (10 * MS_PER_SEC)
+#define SGP30_DELAY_SET_HUMIDITY            (10 * MS_PER_SEC)
+#define SGP30_DELAY_MEASURE_TEST            (220 * MS_PER_SEC)
+#define SGP30_DELAY_GET_FEATURE_SET_VERSION (2 * MS_PER_SEC)
+#define SGP30_DELAY_MEASURE_RAW_SIGNALS     (25 * MS_PER_SEC)
+#define SGP30_DELAY_READ_SERIAL             (500)
+#define SGP30_DELAY_SOFT_RESET              (60 * US_PER_MS)
+/** @} */
+
+/* Device power up time */
+#define SGP30_POWER_UP_TIME         (60 * US_PER_MS)
+/* Polynomial used in crc check */
+#define SGP30_CRC8_POLYNOMIAL       0x31
+/* Seed used in crc check */
+#define SGP30_CRC8_SEED             0xFF
+
+static inline uint8_t _crc8(const void *buf, size_t len)
+{
+    return crc8(buf, len, SGP30_CRC8_POLYNOMIAL, SGP30_CRC8_SEED);
+}
+
+static int _rx_tx_data(sgp30_t *dev, uint16_t cmd, uint8_t *data,
+                       size_t len, uint32_t delay, bool read)
+{
+    int res = 0;
+
+    if (i2c_acquire(dev->params.i2c_dev) != 0) {
+        DEBUG("[sgp30]: could not acquire I2C bus %d\n", dev->params.i2c_dev);
+        return -1;
+    }
+
+    uint8_t frame_cmd[sizeof(cmd) + len];
+    frame_cmd[0] = cmd >> 8;
+    frame_cmd[1] = cmd & 0xFF;
+
+    if (len == 0 || read) {
+        res = i2c_write_bytes(dev->params.i2c_dev, SGP30_I2C_ADDRESS,
+                              &frame_cmd[0], sizeof(cmd), 0);
+    }
+    if (res == 0 && read) {
+        /* delay for command completion */
+        if (!irq_is_in()) {
+            ztimer_sleep(ZTIMER_USEC, delay);
+        }
+        else {
+            ztimer_spin(ZTIMER_USEC, delay);
+        }
+        res = i2c_read_bytes(dev->params.i2c_dev, SGP30_I2C_ADDRESS,
+                             data, len, 0);
+
+    }
+    else if (res == 0 && !read) {
+        memcpy(&frame_cmd[2], data, len);
+        res = i2c_write_bytes(dev->params.i2c_dev, SGP30_I2C_ADDRESS,
+                              frame_cmd, sizeof(cmd) + len, 0);
+        if (res) {
+            DEBUG_PUTS("[sgp30]: fail to write data");
+        }
+    }
+
+    i2c_release(dev->params.i2c_dev);
+
+    return res;
+}
+
+int _read_measurements(sgp30_t *dev, sgp30_data_t *data)
+{
+    uint8_t frame[6];
+
+    if (_rx_tx_data(dev, SGP30_CMD_MEASURE_AIR_QUALITY, frame, sizeof(frame),
+                    SGP30_DELAY_MEASURE_AIR_QUALITY, true)) {
+        return -EPROTO;
+    }
+    if (_crc8(frame, 2) != frame[2] || _crc8(frame + 3, 2) != frame[5]) {
```

Comment:
I've seen this at least two times. maybe factor out a function that does the equivalent of `_crc8(buf, 2) == buf[2]`, then one using that for a full 6 byte frame (`_crc8_matches(frame) && _crc8_matches(frame + 3)`)?